src/Common/Log.cs

```diff
@@ -1,19 +1,40 @@
 ﻿using System;
 using System.IO;
 using System.Text;
+using System.Threading;
 
 namespace Microsoft.SourceBrowser.Common
 {
     public static class Log
     {
-        private static readonly object consoleLock = new object();
         public const string ErrorLogFile = "Errors.txt";
         public const string MessageLogFile = "Messages.txt";
         private const string SeparatorBar = "===============================================";
 
         private static string errorLogFilePath = Path.GetFullPath(ErrorLogFile);
         private static string messageLogFilePath = Path.GetFullPath(MessageLogFile);
 
+        public delegate void MessageHandler(string message, ConsoleColor color);
+        public delegate void FileMessageHandler(string message, string filePath);
+        public static event MessageHandler RaiseMessage;
+        public static event FileMessageHandler RaiseMessageFile;
+
+        private static bool logIsActivated;
+        public static void Activate()
+        {
+            if (logIsActivated)
+                return;
+            var threadLogger = new Thread(() =>
+            {
+                RaiseMessage += InnerWrite;

```

Comment:
Unless I'm mistaken this doesn't work the way you expect. It doesn't matter which thread you're subscribing an event, it will still be raised on the same thread you're calling Invoke from. The thread you create here will just die because it's procedure ends. So effectively this PR is faster because it just removes the lock. You will still see tearing where messages are interrupted halfway and intermingled randomly.

Try maybe using a BlockingCollection: https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1?view=netframework-4.8. The idea to make logging asynchronous is good, because we don't want to be calling File.AppendAllText all the time, it's very wasteful. Should instead open a stream and write to that continuously.

Also consider making an optimization where you only switch the Console color when it's not already what you wanted. Often the color will already be the desired one, in that common case it's best to avoid switching colors unnecessarily. This is confirmed by the trace you're including - switching colors is the most expensive part of the codepath. Worth getting rid of that first.

Reply:
P.S. I'm on vacation until August 12, 2019 so my replies might be delayed. Thanks.